libs/aliengrep/Unit_Match.ml

```diff
@@ -0,0 +1,241 @@
+        failwith ("failed expectation: " ^ show_expectation expectation))
+    expectations
+
+let uconf = Conf.default_uniline_conf
+let mconf = Conf.default_multiline_conf
+
+let test_word () =
+  check uconf {|a|} {|a b c|} [ Num_matches 1; Match_value "a" ];
+  check uconf {|ab|} {|ab abc|} [ Num_matches 1; Match_value "ab" ];
+  check uconf {|ab|} {|ab c ab|} [ Num_matches 2; Match_value "ab" ];
+  check uconf {|ab|} {|xabx|} [ Num_matches 0 ]
+
+let test_whitespace () =
+  check uconf {|a b|} {|a  b|} [ Num_matches 1; Match_value "a  b" ];
+  check uconf {|a  b|} {|a b|} [ Num_matches 1; Match_value "a b" ];
+  check uconf "a\nb" "a b" [ Num_matches 0 ];
+  check uconf "a b" "a\nb" [ Num_matches 0 ];
+  check mconf "a\nb" "a b" [ Num_matches 1; Match_value "a b" ];
+  check mconf "a b" "a\nb" [ Num_matches 1; Match_value "a\nb" ]
+
+let test_ellipsis () =
+  (* test behavior shared with long ellipsis *)
+  check uconf {|a...b|} {|a b|} [ Num_matches 1; Match_value "a b" ];
+  check uconf {|a...b|} {|a/b|} [ Num_matches 1; Match_value "a/b" ];
+  check uconf {|a...b|} {|a x y b|} [ Num_matches 1; Match_value "a x y b" ];
+  check uconf {|a...b|} {|a b b|} [ Num_matches 1; Match_value "a b" ];
+  check uconf {|a...b b c|} {|a b b b c|}
+    [ Num_matches 1; Match_value "a b b b c" ];
+  (* test behavior specific to regular ellipsis *)
+  check uconf {|a...b|} "a\nb" [ Num_matches 0 ];
+  (* in multiline mode, a regular ellipsis matches newlines *)
+  check mconf {|a...b|} "a\nx\nx\nb" [ Num_matches 1; Match_value "a\nx\nx\nb" ]
+
+let test_long_ellipsis () =
+  (* test behavior shared with regular ellipsis *)
+  check uconf {|a....b|} {|a b|} [ Num_matches 1; Match_value "a b" ];
+  check uconf {|a....b|} {|a/b|} [ Num_matches 1; Match_value "a/b" ];
+  check uconf {|a....b|} {|a x y b|} [ Num_matches 1; Match_value "a x y b" ];
+  check uconf {|a....b|} {|a b b|} [ Num_matches 1; Match_value "a b" ];
+  check uconf {|a....b b c|} {|a b b b c|}
+    [ Num_matches 1; Match_value "a b b b c" ];
+  (* test behavior specific to long ellipsis *)
+  check uconf {|a....b|} "a\nb" [ Num_matches 1; Match_value "a\nb" ];
+  check mconf {|a....b|} "a\nx\nx\nb"
+    [ Num_matches 1; Match_value "a\nx\nx\nb" ]
+
+let test_metavariables () =
+  check uconf {|a $X b|} {|a xy b|}
+    [
+      Num_matches 1;
+      Match_value "a xy b";
+      Capture (Metavariable, "X");
+      Capture_value ((Metavariable, "X"), "xy");
+    ];
+  check uconf {|a $AB_4! b|} {|a xy! b|}
+    [
+      Num_matches 1;
+      Match_value "a xy! b";
+      Capture_value ((Metavariable, "AB_4"), "xy");
+    ];
+  check uconf {|$ X|} {|$ X|}
+    [
+      Num_matches 1;
+      Match_value "$ X";
+      Not (Capture (Metavariable, "X"));
+      Not (Capture (Metavariable, ""));
+      Not (Capture (Metavariable, "$"));
+    ];
+  check uconf {|$A $B|} {|1 2 3 4|}
+    [
+      (* at the moment, matches don't overlap -> no match on "2 3" *)
```

Comment:
I believe it's the OCaml `Pcre` search function that does that rather than a C function from the original PCRE library. If true, we may be able to write our own search function that allows overlaps.